### DISCUSSION TITLE
Problem getting basic simple build script to work

### DISCUSSION THREAD

**Discussion by BillHertzing** [nightroman/Invoke-Build#207] [General]
I'd like to use Invoke-Build as the core build script for the powershell modules in my public MIT-licensed GitHub Repository, which is a large monorepo with multiple Powershell modules. But I can't get a basic simple build script to run.  The simple script is below. It has an Enter-Build scriptblock, and three simple tasks (they just log the task entry). The output says that two of the tasks are redeneid. I can't figure out why.  I must be making some kind of simple mistake, but I've spent a few days now with no success.

if it matters, most of my scripts in my repro  DO expect/use $PSDefaultParameterValues as follows:
```Powershell
Name                           Value
----                           -----
*:Encoding                     UTF8
```

_invoke-build_ produces the following output:
 .\ATAP.Utilities.BuildTooling.PowerShell\ > invoke-build -ModuleRoot .    
Build . <obfuscated>\GitHub\ATAP.Utilities\src\ATAP.Utilities.BuildTooling.PowerShell\module.build.ps1
Redefined task 'BuildPSM1'.
Redefined task 'BuildPSD1'.
Task /./BuildPSD1/BuildPSM1/Clean
Done /./BuildPSD1/BuildPSM1/Clean 00:00:00.0030836
Done /./BuildPSD1/BuildPSM1 00:00:00.0034365
Done /./BuildPSD1 00:00:00.0036953
Done /. 00:00:00.0039474
Build succeeded. 4 tasks, 0 errors, 0 warnings 00:00:00.0289462

_invoke-build_ with the _-debug_ flag produces the following output:
invoke-build -ModuleRoot . -debug
Build . <obfuscated>\GitHub\ATAP.Utilities\src\ATAP.Utilities.BuildTooling.PowerShell\module.build.ps1
Redefined task 'BuildPSD1'.
Redefined task 'BuildPSM1'.
DEBUG: 19 | [22:30:30][<ScriptBlock>] Starting Module.Build.ps1; Configuration = Production; BuildRoot = <obfuscated>\GitHub\ATAP.Utilities\src\ATAP.Utilities.BuildTooling.PowerShell; Encoding = UTF8
DEBUG: 39 | [22:30:30][<ScriptBlock>] ModuleRoot = .; moduleName = ATAP.Utilities.BuildTooling.PowerShell
DEBUG: 40 | [22:30:30][<ScriptBlock>] sourceFileInfos = <large number of (correct) full file paths>
DEBUG: 41 | [22:30:30][<ScriptBlock>] sourceManifestPath = .\ATAP.Utilities.BuildTooling.PowerShell.psd1; GeneratedModuleFilePath = .\_generated\ATAP.Utilities.BuildTooling.PowerShell.psm1; GeneratedManifestFilePath = .\_generated\ATAP.Utilities.BuildTooling.PowerShell.psd1;
Task /./BuildPSD1/BuildPSM1/Clean
DEBUG: 49 | [22:30:30][<ScriptBlock>] Starting Task Clean; Configuration = Production; BuildRoot = <obfuscated>\GitHub\ATAP.Utilities\src\ATAP.Utilities.BuildTooling.PowerShell; Encoding = UTF8
DEBUG: 50 | [22:30:30][<ScriptBlock>] OriginalLocation = <obfuscated>\GitHub\ATAP.Utilities\src\ATAP.Utilities.BuildTooling.PowerShell; BuildFile = <obfuscated>\GitHub\ATAP.Utilities\src\ATAP.Utilities.BuildTooling.PowerShell\module.build.ps1
DEBUG: 51 | [22:30:30][<ScriptBlock>] WhatIf:False; -Verbose:SilentlyContinue; -Confirm:High
Done /./BuildPSD1/BuildPSM1/Clean 00:00:00.0033031
Done /./BuildPSD1/BuildPSM1 00:00:00.0035807
Done /./BuildPSD1 00:00:00.0038189
Done /. 00:00:00.0040739
Build succeeded. 4 tasks, 0 errors, 0 warnings 00:00:00.0294895

This shows that the 'Enter-Build' script and the 'Clean' task both ran, but the other two tasks did not run.

Help, Please?

 I am using the latest versions of VSC (1.73.1), Powershell (7.3.0), Powershell VSC Extension (11/19/2022), and Invoke-Build (5.10.1) on Windows 10.0.22000 (Windows 11)

 Module.Build,ps1:
```Powershell

param(

  [parameter(ValueFromPipeline = $false, ValueFromPipelineByPropertyName = $True, Mandatory = $True)]
  [ValidateNotNullOrEmpty()]
  [string]  $ModuleRoot
  ,[ValidateSet('Development', 'QualityAssurance', 'Production')]
  [string] $Configuration = 'Production'
  , [string] $BuildRoot = $BuildRoot
  , [parameter(ValueFromPipeline = $false, ValueFromPipelineByPropertyName = $True, Mandatory = $False)]
  [string] $Encoding # Often found in the $PSDefaultParameterValues preference variable
)

Enter-Build {
# Write-PSFMessage will write to its log file regardless of the value of $DebugPreference
Write-PSFMessage -Level Debug -Message "Starting Module.Build.ps1; Configuration = $Configuration; BuildRoot = $BuildRoot; Encoding = $Encoding"

# By convention the module name and the final part of the $moduleRoot path are the same
$moduleName = Split-Path $ModuleRoot -Leaf
$moduleFilename =  $moduleName + '.psm1'
$manifestFilename = $moduleName + '.psd1'

$sourceSubDirectorys = @('public', 'private', 'Resources')
$sourceExtensions = @('.ps1', '.clixml')
$sourceFileInfos = $sourceSubDirectorys | ForEach-Object { $sourceSubDirectory = $_; $sourceExtensions | ForEach-Object { $sExtension = $_;
    (Get-ChildItem $(Join-Path $sourceSubDirectory $('*' + $sExtension)))
  } }

$sourceManifestPath = Join-Path $moduleRoot $manifestFilename

$GeneratedPowerShellModuleSubdirectory = '_generated'
$GeneratedModuleFilePath = Join-Path '.' $GeneratedPowerShellModuleSubdirectory $moduleFilename
$GeneratedManifestFilePath = Join-Path '.' $GeneratedPowerShellModuleSubdirectory $manifestFilename

Write-PSFMessage -Level Debug -Message "ModuleRoot = $ModuleRoot; moduleName = $moduleName"
Write-PSFMessage -Level Debug -Message "sourceFileInfos = $sourceFileInfos"
Write-PSFMessage -Level Debug -Message "sourceManifestPath = $sourceManifestPath; GeneratedModuleFilePath = $GeneratedModuleFilePath; GeneratedManifestFilePath = $GeneratedManifestFilePath;"

## Validate or create required output paths
if (-not $(Test-Path -Path $GeneratedPowerShellModuleSubdirectory -PathType Container)) { $null = New-Item -ItemType Directory -Force $GeneratedPowerShellModuleSubdirectory }

}

Task Clean {
  Write-PSFMessage -Level Debug -Message "Starting Task Clean; Configuration = $Configuration; BuildRoot = $BuildRoot; Encoding = $Encoding"
  Write-PSFMessage -Level Debug -Message "OriginalLocation = $OriginalLocation; BuildFile = $BuildFile"
  Write-PSFMessage -Level Debug -Message "WhatIf:$WhatIfPreference; -Verbose:$Verbosepreference; -Confirm:$ConfirmPreference"
}

Task BuildPSM1 -Inputs $sourceFileInfos -Outputs $GeneratedModuleFilePath {
  Write-PSFMessage -Level Debug -Message "Starting Task BuildPSM1; GeneratedModuleFilePath = $GeneratedModuleFilePath; inputs = $inputs"
}

Task BuildPSD1 -inputs $sourceFileInfos -Outputs $GeneratedManifestFilePath {
  Write-PSFMessage -Level Important -Message "Starting Task BuildPSD1; GeneratedManifestFilePath = $GeneratedManifestFilePath; inputs = $inputs"
}

# Default task.
Task . BuildPSD1

Task BuildPSD1 BuildPSM1

Task BuildPSM1 Clean

```

**Comment by nightroman** [comment]
As the information message tells, you have redefined tasks:

> Redefined task 'BuildPSD1'.

```powershell
Task BuildPSD1 -inputs $sourceFileInfos -Outputs $GeneratedManifestFilePath { ...
Task BuildPSD1 BuildPSM1
```

> Redefined task 'BuildPSM1'.

```powershell
Task BuildPSM1 -Inputs $sourceFileInfos -Outputs $GeneratedModuleFilePath { ...
Task BuildPSM1 Clean
```

This is not a build error as such because in rare cases an author wants to redefine tasks.
The last defined/added task with the same name wins, the others are gone.
MSBuild works in the same way with targets.
PowerShell works in the same way with functions.

Thus, unless you deliberately redefine tasks (unlikely per the code), it's the bug.

Remove these lines or use different task names:

```powershell
Task BuildPSD1 BuildPSM1
Task BuildPSM1 Clean
```

**Comment by BillHertzing** [comment]
Ah! As expected, a simple misunderstanding on my part. I'm past the first hurdle, now on to writing more complex tasks. Thanks for your help!
